File: ios/MenCare/Core/Model/Operator.swift
import Foundation

struct TimeRange: Hashable, Comparable {
    let start: LocalTime
    let end: LocalTime

    init(_ start: LocalTime, _ end: LocalTime) {
        precondition(start < end, "start must be before end")
        self.start = start
        self.end = end
    }

    func overlaps(_ other: TimeRange) -> Bool { start < other.end && other.start < end }

    /// Overlapping part of the two ranges, or nil when they don't overlap.
    func intersect(_ other: TimeRange) -> TimeRange? {
        let from = max(start, other.start)
        let to = min(end, other.end)
        return from < to ? TimeRange(from, to) : nil
    }

    static func < (lhs: TimeRange, rhs: TimeRange) -> Bool { lhs.start < rhs.start }
}

struct Operator: Identifiable, Hashable {
    let id: String
    var name: String
    var title: String
    var bio: String
    var specialties: [String]
    var isOwner: Bool = false
    var weeklyHours: [DayOfWeek: [TimeRange]]
    var serviceIds: Set<String>

    var initials: String {
        name.split(separator: " ").prefix(2).compactMap { $0.first.map(String.init)?.uppercased() }.joined()
    }
}

enum BlockReason: CaseIterable, Hashable {
    case permesso, pausa, ferie, corso
}

struct TimeBlock: Identifiable, Hashable {
    let id: String
    let operatorId: String
    var reason: BlockReason
    var date: LocalDate
    var range: TimeRange
    var label: String?
}

struct Holiday: Identifiable, Hashable {
    let id: String
    let operatorId: String
    var from: LocalDate
    var to: LocalDate
    var label: String
}

File: android/core/model/src/main/kotlin/com/devora/mencare/core/model/Operator.kt
package com.devora.mencare.core.model

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalTime

data class TimeRange(
    val start: LocalTime,
    val end: LocalTime,
) {
    init {
        require(start < end) { "start must be before end" }
    }

    fun overlaps(other: TimeRange): Boolean = start < other.end && other.start < end

    /** Overlapping part of the two ranges, or null when they don't overlap. */
    fun intersect(other: TimeRange): TimeRange? {
        val from = maxOf(start, other.start)
        val to = minOf(end, other.end)
        return if (from < to) TimeRange(from, to) else null
    }
}

data class Operator(
    val id: String,
    val name: String,
    val title: String,
    val bio: String,
    val specialties: List<String>,
    val isOwner: Boolean = false,
    val weeklyHours: Map<DayOfWeek, List<TimeRange>>,
    val serviceIds: Set<String>,
) {
    val initials: String
        get() = name.split(' ').filter { it.isNotBlank() }.take(2)
            .joinToString("") { it.first().uppercase() }
}

enum class BlockReason { PERMESSO, PAUSA, FERIE, CORSO }

data class TimeBlock(
    val id: String,
    val operatorId: String,
    val reason: BlockReason,
    val date: LocalDate,
    val range: TimeRange,
    val label: String? = null,
)

data class Holiday(
    val id: String,
    val operatorId: String,
    val from: LocalDate,
    val to: LocalDate,
    val label: String,
)
